core/src/apps/management/recovery_device/__init__.py

```diff
@@ -39,16 +39,24 @@ async def recovery_device(msg: RecoveryDevice) -> Success:
 
     # --------------------------------------------------------
     # validate
-    if recovery_type == RecoveryType.NormalRecovery:
+    if recovery_type is RecoveryType.NormalRecovery:
         if storage_device.is_initialized():
             raise wire.UnexpectedMessage("Already initialized")
-    elif recovery_type in (RecoveryType.DryRun, RecoveryType.UnlockRepeatedBackup):
+    elif recovery_type is RecoveryType.DryRun:
+        if not storage_device.is_initialized():
+            raise wire.NotInitialized("Device is not initialized")
+        elif storage_device.no_backup():
+            raise wire.ProcessError("Dry-run not available for seedless devices")
+        elif storage_device.needs_backup() or storage_device.unfinished_backup():
+            raise wire.ProcessError("Cannot do dry-run without backed-up seed")
+        # check that only allowed fields are set
+        for key, value in msg.__dict__.items():
+            if key not in DRY_RUN_ALLOWED_FIELDS and value is not None:
+                raise wire.ProcessError(f"Forbidden field set in dry-run: {key}")
+    elif recovery_type is RecoveryType.UnlockRepeatedBackup:
         if not storage_device.is_initialized():
             raise wire.NotInitialized("Device is not initialized")
-        if (
```

Comment:
nit: can we add the following `if` here? IIUC, it should help to avoid `DryRun` & `UnlockRepeatedBackup` checks' duplication:
```python
        if recovery_type is RecoveryType.DryRun:
            if storage_device.no_backup():
                raise wire.ProcessError("Dry-run not available for seedless devices")
            elif storage_device.needs_backup() or storage_device.unfinished_backup():
                raise wire.ProcessError("Cannot do dry-run without backed-up seed")
```